tests/project_helpers.py

```diff
@@ -0,0 +1,238 @@
+from dataclasses import dataclass
+from dataclasses import fields as dataclass_fields
+from pathlib import Path
+from typing import Any, Callable, Dict, Iterable, Iterator, List, Optional, Set, Type
+
+import pytest
+
+from fawltydeps.main import Analysis
+from fawltydeps.types import TomlData
+
+if sys.version_info >= (3, 11):
+    import tomllib  # pylint: disable=E1101
+else:
+    import tomli as tomllib
+
+JsonData = Dict[str, Any]
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class CachedExperimentVenv:
+    """A virtualenv used in a sample or real-world experiment.
+
+    This is a test helper to create and reuse virtualenvs that contain real
+    packages used in our integration tests. The goal is that venvs are created
+    _once_ and reused as long as the commands used to create them and install
+    packages into them remain unchanged. These use pytest's caching
+    infrastructure meaning that the venvs typically end up living under
+    ~/.cache/pytest/d/...
+    """
+
+    requirements: List[str]  # PEP 508 requirements, passed to 'pip install'
+
+    def venv_script_lines(self, venv_path: Path) -> List[str]:
+        return (
+            [
+                f"rm -rf {venv_path}",
+                f"python3 -m venv {venv_path}",
+                f"{venv_path}/bin/pip install --upgrade pip",
+            ]
+            + [
+                f"{venv_path}/bin/pip install {shlex.quote(req)}"
+                for req in self.requirements
+            ]
+            + [
+                f"touch {venv_path}/.installed",
+            ]
+        )
+
+    def venv_hash(self) -> str:
+        """Returns a hash that depends on the venv script and python version.
+
+        The installation script will change if the code to setup the venv in
+        venv_script_lines() changes, or if the requirements of the experiment
+        changes. It will also be different for different Python versions.
+        The Python version currently used to run the tests is used to compute
+        the hash and create the venv.
+        """
+        dummy_script = self.venv_script_lines(Path("/dev/null"))
+        py_version = f"{sys.version_info.major},{sys.version_info.major}"
+        script_and_version_bytes = ("".join(dummy_script) + py_version).encode()
+        return hashlib.sha256(script_and_version_bytes).hexdigest()
+
+    def __call__(self, cache: pytest.Cache) -> Path:
+        """Get this venv's dir and create it if necessary.
+
+        The venv_dir is where we install the dependencies of the current
+        experiment. It is keyed by the sha256 checksum of the requirements
+        file and the script we use for setting up the venv. This way, we
+        don't risk using a previously cached venv for a different if the
+        script or the requirements to create that venv change.
+        """
+        # We cache venv dirs using the hash from create_venv_hash
+        cached_str = cache.get(f"fawltydeps/{self.venv_hash()}", None)
```

Comment:
Nit: should be able to drop the `None` here, right? I think that's the default.

Reply:
Actually, no, this is pytest's `Cache.get()`m and neither the [docs](https://docs.pytest.org/en/7.1.x/reference/reference.html?highlight=cache%20get#pytest.Cache.get) nor the [source itself](https://docs.pytest.org/en/7.1.x/_modules/_pytest/cacheprovider.html#Cache.get) provide a `None` default-value for `default`, so we must pass it ourselves.